average_power.c

```diff
@@ -0,0 +1,74 @@
+#include <altivec.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/time.h>
+
+// Take the number 1, treat it as long integer and shift its binary
+// representation 30 positions to the left => N = 2^32
+#define N (1L << 30)
+
+int main() {
+
+  // allocate memory for array
+  // A array of size N
+  float *A = NULL;
+  if (posix_memalign((void **)&A, 16, N * sizeof(float)) != 0) {
+    perror("posix_memalign failed");
+    exit(EXIT_FAILURE);
+  }
+  // float *A = aligned_alloc(16, N * sizeof(float)); (maybe use this instead)
+
+  // fill array with numbers
+  for (int i = 0; i < N; i++) {
+    A[i] = (float)i;
+  }
+
+  float avg1 = 0.0, avg2 = 0.0;
+  struct timeval tv1, tv2, tv3, diff1, diff2;
+  gettimeofday(&tv1, NULL);
+
+  // SCALAR VERSION
+  // calculates average in avg1 variable
+  for (int i = 0; i < N; i++) {
+    avg1 += A[i];
+  }
+  avg1 /= (float)N;
+  gettimeofday(&tv2, NULL);
+
+  // POWER VERSION (george mermigkis)
+  // declare vector that includes 4 floats & initialize them with val=0
+  vector float vec = {0.0f, 0.0f, 0.0f, 0.0f};
+
+  // do N/4 iterations because every vector will take 4 loops each time
+  for (int i = 0; i < N; i += 4) {
+
+    // load the corresponding 4 values from the loop, perform addition & store
+    // result in vec
+    vector float input = vec_ld(0, &A[i]);
+    vec = vec_add(vec, input);
+  }
+
+  // add all the 4 parts of the vector and save them in sum vector
+  float sum_array[4];
+  vec_st(vec, 0, sum_array);
+  avg2 = sum_array[0] + sum_array[1] + sum_array[2] + sum_array[3];
```

Comment:
there is a way to do it in VSX before the vec_st, in essence it's this:
// vec = | V1 | V2 | V3 | V4 |
tmp = vec_sld(vec, vec, 8); // shift vector 8 bytes to the left tmp | V3 | V4 | V1 | V2 |
tmp = vec_add(vec, tmp); // add vec + tmp -> | V3 + V1 | V4 + V2 | V1 + V3 | V2 + V4 |
tmp2 = vec_sld(tmp,tmp, 4); // tmp2: | V4 + V2 | V1 + V3 | V2 + V4 | V3 + V1 |
vec = vec_add(tmp, tmp2); // vec = | V3 + V1 + V4 + V2 | ...
vec_st(...) // all elements have the same value
// or vec_ste() single element 0
return sum_array[0].